### PR TITLE
Audio query functionality to multimodal backend

### PR DESCRIPTION
**Pull request by okhleif-10**
## Description

This PR enables audio as a query in the backend, where the expected input to the gateway is a b64 string.

## Issues

Part of RFC --> https://github.com/opea-project/docs/blob/main/community/rfcs/24-10-02-GenAIExamples-001-Image_and_Audio_Support_in_MultimodalQnA.md

## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)
- [ ] Others (enhancement, documentation, validation, etc.)

## Dependencies

N/A

## Tests

Verify with curl command:

`curl http://${host_ip}:8888/v1/multimodalqna     -H "Content-Type: application/json"     -d '{"messages": [{"role": "user", "content": [{"type": "audio", "audio": "UklGRigAAABXQVZFZm10IBIAAAABAAEARKwAAIhYAQACABAAAABkYXRhAgAAAAEA"}]}]}'`

`curl http://${host_ip}:8888/v1/multimodalqna     -H "Content-Type: application/json"     -d '{"messages": [{"role": "user", "content": [{"type": "audio", "audio": "UklGRigAAABXQVZFZm10IBIAAAABAAEARKwAAIhYAQACABAAAABkYXRhAgAAAAEA, "}]}, {"role": "assistant", "content": "How may I help?"}, {"role": "user", "content": [{"type": "text", "text": "what pronoun did I tell you?"}]}]}`

Need to add automated test.

### CODE REVIEW

File: comps/embeddings/multimodal/multimodal_langchain/mm_embedding_mmei.py [hunk 1]
```diff
@@ -9,6 +9,7 @@
 from fastapi.responses import JSONResponse
 
 from comps import (
+    Base64ByteStrDoc,
```
Comment (dmsuehir) [inline comps/embeddings/multimodal/multimodal_langchain/mm_embedding_mmei.py]: Do we still need the changes in this file? I'm thinking that by the time the prompt gets to the embedding service, it should be text (not audio/b64), so these changes can probably be reverted?

File: comps/cores/proto/docarray.py [hunk 2]
```diff
@@ -63,17 +63,16 @@ class TextImageDoc(BaseDoc):
     text: TextDoc = None
 
 
+class Base64ByteStrDoc(BaseDoc):
+    byte_str: str
+
 MultimodalDoc = Union[
+    Base64ByteStrDoc,
```
Comment (dmsuehir) [inline comps/cores/proto/docarray.py]: I don't think we need this any more. We can move the `Base64ByteStrDoc` back to where it used be after `MultimodalDoc`

**Review by dmsuehir — CHANGES_REQUESTED** [review]

File: comps/cores/mega/gateway.py [hunk 3]
```diff
@@ -859,16 +880,47 @@ async def handle_request(self, request: Request):
...
+        messages = self._handle_message(chat_request.messages)
+        if isinstance(messages, tuple):
             # print(f"This request include image, thus it is a follow-up query. Using lvm megaservice")
-            prompt, images = prompt_and_image
+            prompt, b64_types = messages
+            decoded_audio_input = ""
+            if "audio" in b64_types:
+                # call ASR endpoint to decode audio to text             
+                input_dict = {"byte_str": b64_types["audio"][0]}
+                response = requests.post(self.asr_endpoint, data=json.dumps(input_dict), proxies={"http": None})
```
Comment (dmsuehir) [inline comps/cores/mega/gateway.py]: Since we are doing this same thing in a couple of places, maybe create an audio to text function to reduce duplication.

**Review by dmsuehir — CHANGES_REQUESTED** [review]

File: comps/cores/proto/docarray.py (hunk 2 above)
Comment (okhleif-10) [reply comps/cores/proto/docarray.py]: So just remove line 70?

File: comps/embeddings/multimodal/multimodal_langchain/mm_embedding_mmei.py (hunk 1 above)
Comment (okhleif-10) [reply comps/embeddings/multimodal/multimodal_langchain/mm_embedding_mmei.py]: Probably not, this was from your branch and I didn't look too much into it

File: comps/cores/proto/docarray.py (hunk 2 above)
Comment (dmsuehir) [reply comps/cores/proto/docarray.py]: Line 70 and The `Base64ByteStrDoc` can go back to where it was before. I only moved it because I was referencing it from `MultimodalDoc`, so it had to be declared before that.

File: comps/embeddings/multimodal/multimodal_langchain/mm_embedding_mmei.py (hunk 1 above)
Comment (dmsuehir) [reply comps/embeddings/multimodal/multimodal_langchain/mm_embedding_mmei.py]: I think we can revert it

**Review by mhbuehler — COMMENTED** [review]
@okhleif-IL thank you. This LGMT but I need to continue testing it with real audio. Also, it might need to be rebased onto a new branch now that `mm-rag-enhancements` has been upstreamed as Phase 1.

**Review by dmsuehir — COMMENTED** [review]
I tried out the branch, and think there's an issue with the follow up query using audio where it's going through the full flow of going back to the embedding/retriever/vector store vs just to the LVM. I thought that we saw earlier that follow up queries were skipping embedding, etc and going straight to the LVM. Adding audio support shouldn't change that (it essentially should work the same as if the query was text, except that the audio needs to be translated to text)

**Review by mhbuehler — COMMENTED** [review]
Are there any meaningful changes to `comps/cores/proto/docarray.py` or `comps/embeddings/multimodal/multimodal_langchain/mm_embedding_mmei.py`? If not, I'd leave them out of the PR.

File: comps/cores/mega/gateway.py [hunk 4]
```diff
@@ -815,8 +819,13 @@ def _handle_message(self, messages):
                         image_list = [
                             item["image_url"]["url"] for item in message["content"] if item["type"] == "image_url"
                         ]
+                        audios = [
+                            item["audio"] for item in message["content"] if item["type"] == "audio"
+                        ]
                         if image_list:
                             messages_dict[msg_role] = (text, image_list)
+                        elif audios:
```
Comment (mhbuehler) [inline comps/cores/mega/gateway.py]: We probably want to allow a message to contain both an image AND an audio, for example if it's a follow-up query about an image but the query is in audio format.

**Review by mhbuehler — CHANGES_REQUESTED** [review]

File: comps/cores/mega/gateway.py [hunk 5]
```diff
@@ -985,18 +1044,25 @@ async def handle_request(self, request: Request):
...
             if "detail" in result_dict[last_node].keys():
                 response = result_dict[last_node]["detail"]
             else:
-                response = "The server fail to generate answer to your query!"
+                response = "The server failed to generate an answer to your query!"
         if "metadata" in result_dict[last_node].keys():
             # from retrieval results
-            metadata = result_dict[last_node]["metadata"]
+            if decoded_audio_input:
+                metadata = {"audio": decoded_audio_input}
```
Comment (mhbuehler) [inline comps/cores/mega/gateway.py]: The decoded audio should be added to the existing metadata, not replace it. I am not seeing the retrieved image data in the response, so please change this to: `metadata["audio"] = decoded_audio_input`

**Review by mhbuehler — CHANGES_REQUESTED** [review]

File: comps/cores/mega/gateway.py [hunk 6]
```diff
@@ -1053,7 +1053,7 @@ async def handle_request(self, request: Request):
         if "metadata" in result_dict[last_node].keys():
             # from retrieval results
             if decoded_audio_input:
-                metadata = {"audio": decoded_audio_input}
+                metadata["audio"] =  decoded_audio_input
```
Comment (mhbuehler) [inline comps/cores/mega/gateway.py]: Oh, this doesn't work because metadata doesn't exist yet. So maybe:
```
if "metadata" in result_dict[last_node].keys():
    metadata = result_dict[last_node]["metadata"]
    if decoded_audio_input:
        metadata["audio"] =  decoded_audio_input    
```

**Review by mhbuehler — CHANGES_REQUESTED** [review]

**Review by mhbuehler — APPROVED** [review]

File: comps/cores/mega/gateway.py [hunk 7]
```diff
@@ -943,15 +990,26 @@ async def handle_request(self, request: Request):
...
+                initial_inputs = {"prompt": prompt, "image": b64_types["image"][0]}
+            else:    
+                # This query includes only text, translated audio, or both, hence it will use Embedding
+                cur_megaservice = self.megaservice
+                initial_inputs = {"text": prompt}
+
         else:
-            # print(f"This is the first query, requiring multimodal retrieval. Using multimodal rag megaservice")
-            prompt = prompt_and_image
+            # This is a query with only text, hence it is a first query and is using Embedding
```
Comment (dmsuehir) [inline comps/cores/mega/gateway.py]: We can't assume that because it's text only it's a first query. The first query could have matched with something like a podcast clip (which is text only), so follow up queries will still be text only. I had a similar issue when implementing image queries and here's how I'm solving it on [my branch](https://github.com/mhbuehler/GenAIComps/blob/dina/image_query/comps/cores/mega/gateway.py#L971):
```
num_messages = len(data["messages"]) if isinstance(data["messages"], list) else 1
```

**Review by dmsuehir — CHANGES_REQUESTED** [review]

File: comps/cores/mega/gateway.py [hunk 8]
```diff
@@ -865,16 +870,28 @@ def _handle_message(self, messages):
                     system_prompt = message["content"]
                 elif msg_role == "user":
                     if type(message["content"]) == list:
+                         # separate each media type and store accordingly
```
Comment (dmsuehir) [inline comps/cores/mega/gateway.py]: Indentation is off
```suggestion
                        # separate each media type and store accordingly
```

File: comps/cores/mega/gateway.py [hunk 9]
```diff
@@ -889,51 +906,81 @@ def _handle_message(self, messages):
...
+
+        response = requests.post(self.asr_endpoint, data=json.dumps(input_dict), proxies={"http": None})
+            
+        if response.status_code != 200:
+            return JSONResponse(status_code=503, content={"message": "Unable to convert audio to text. {}".format(
+                response.text)})
+
+        response = response.json()
+        return response["query"]
+
```
Comment (dmsuehir) [inline comps/cores/mega/gateway.py:983]: For consistency, just have one empty line between the functions

File: comps/cores/mega/gateway.py [hunk 10]
```diff
@@ -966,6 +1032,7 @@ async def handle_request(self, request: Request):
             streaming=stream_opt,
             chat_template=chat_request.chat_template if chat_request.chat_template else None,
         )
+
```
Comment (dmsuehir) [inline comps/cores/mega/gateway.py]: This empty line can be removed so that we don't have an unnecessary diff

**Review by dmsuehir — CHANGES_REQUESTED** [review]

**Review by dmsuehir — APPROVED** [review]
LGTM, thanks
